### PR TITLE
Introduce concrete use cases for navigating and post-processing dependencies

### PR DESCRIPTION
**Pull request by bmuschko**
### Context

For more information see https://github.com/gradle/dotorg-docs/issues/170.

__Additional information:__

- I removed the existing content. The examples weren't really relatable. The text didn't indicate a use case.
- I decided not to move the content under "Inspecting dependencies". There are use cases that don't quite fit to the title. For now I left it underneath "Working with Dependencies". We can come up with a better name though I haven't found a good one yet. Maybe "Navigating dependencies and resolved modules"?
- For now I picked three common scenarios. We can certainly enhance the section by more use cases in the future.
- Please let me know if the use of a particular API is "deprecated" or "silently deprecated" and should not be used in the user manual for reason x.

### CODE REVIEW

File: subprojects/docs/src/samples/userguide/dependencies/iteratingArtifacts/build.gradle [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+dependencies {
+    scm 'org.eclipse.jgit:org.eclipse.jgit:4.9.2.201712150930-r'
+    scm 'commons-codec:commons-codec:1.7'
+}
+
+//START SNIPPET iteration-task
+task iterateScmArtifacts {
+    doLast {
```
Comment (big-guy) [inline subprojects/docs/src/samples/userguide/dependencies/iteratingArtifacts/build.gradle]: These sort of tasks need to have dependencies on the configuration or else composite builds may not work.

Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencies/iteratingArtifacts/build.gradle]: Thanks for the info. I think that's important information to have on the page on composite builds. I will change the code as such:

```
task iterateScmArtifacts {
     dependsOn configurations.scm
     ...
}
```

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 2]
```diff
@@ -15,73 +15,47 @@
...
-        </sample>
-++++
-
-The dependencies have the following transitive dependencies:
-
-shark-1.0 -&gt; seal-2.0, tuna-1.0
+=== Iterating over dependencies assigned to a configuration
 
-orca-1.0 -&gt; seal-1.0
+The <<sec:listing_dependencies,dependencies report>> is a built-in task that comes with every project independent of the applied plugins. Sometimes you'll want to implement logic based on the resolved dependencies of a project e.g. for building a plugin that visualizes the dependency graph as an image. The api:org.gradle.api.artifacts.Configuration[] API provides a wealth of methods to navigate and inspect dependencies assigned to a configuration.
```
Comment (jjohannes) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: I see how this relates to the _dependency report_ as you can use the API for similar things. But as the first sentences (_The dependencies report is a built-in task that comes with every project independent of the applied plugins._), it feels out of place to me. Maybe just remove it?
Or add it as a not at the end of the section below the example. Would also give an opportunity to mention build scans. Something like:
_Note that if you only want to analyses dependencies of your build, you do not need to write any custom script code. You can use build scans(LINK) or the built-in dependencies(LINK) task._

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 3]
```diff
@@ -15,73 +15,47 @@
...
-To get the library files of the configuration dependencies you can do:
+=== Iterating over artifacts resolved for a module
 
-++++
-<sample id="configurationHandlingAllFiles" dir="userguide/artifacts/configurationHandling" title="Configuration.files">
-            <sourcefile file="build.gradle" snippet="allFiles"/>
-            <output args="-q allFiles"/>
-        </sample>
-++++
+None of the <<inspecting_dependencies,dependency reporting>> helps you with inspecting or further processing the underlying, resolved artifacts of a module. A typical use case for wanting to access the artifacts is to copy them into a specific directory or filter out files of interest based on a specific file extension.
```
Comment (jjohannes) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: "use case for wanting to access the artifacts" -> "use case for accessing the artifacts"

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 4]
```diff
@@ -15,73 +15,47 @@
...
 ++++
 
-The `Configuration.files` method always retrieves all artifacts of the _whole_ configuration. It then filters the retrieved files by specified dependencies. As you can see in the example, transitive dependencies are included.
+[NOTE]
+====
+Iterating over the artifacts of a module automatically resolves the configuration. A resolved configuration becomes immutable and cannot add or remove dependencies. If needed you can copy a configuration for further modification via api:org.gradle.api.artifacts.Configuration#copy()[].
+====
 
-You can also copy a configuration. You can optionally specify that only a subset of dependencies from the original configuration should be copied. The copying methods come in two flavors. The `copy` method copies only the dependencies belonging explicitly to the configuration. The `copyRecursive` method copies all the dependencies, including the dependencies from extended configurations.
+=== Accessing a module's metadata artifact
```
Comment (jjohannes) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: "Accessing a module’s metadata artifact" -> "Accessing a module’s metadata file"

We should not talk about _metada artifacts_ as an artifact is something specific for us. Talk about _metadata_ or _metadata files_. I know that we have `MavenPomArtifact` in the API. Still I would avoid the term in the text.

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 5]
```diff
@@ -15,73 +15,47 @@
...
 
-++++
-<sample id="configurationHandlingCopy" dir="userguide/artifacts/configurationHandling" title="Configuration.copy">
-            <sourcefile file="build.gradle" snippet="copy"/>
-            <output args="-q copy"/>
-        </sample>
-++++
+As part of the dependency resolution process, Gradle downloads the main artifact and the metadata file for a module and stores them in the dependency cache. The initial download requires a active connection to the repositories declared in the build. Some organizations enforce strong restrictions on accessing repositories outside of internal network. Instead of downloading artifacts, those organizations prefer to provide an "installable" Gradle cache with all artifacts contained in it to fulfill the build's dependency requirements.
```
Comment (jjohannes) [inline subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: :x: "Gradle downloads the main artifact and the metadata file for a module and stores them" -> "Gradle downloads the metadata file of a module and stores it"

(If the "main artifact" is not always downloaded for every dependency. Depends on repository type and the new metadata sources configuration.)

Comment (jjohannes) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: The part after the first sentence starting with _The initial download..._  feels unrelated to this section. Which is only about the modification api, no? Maybe there is some better place in the docs for this?

File: subprojects/docs/src/samples/userguide/dependencies/iteratingArtifacts/build.gradle (hunk 1 above)
Comment (bmuschko) [reply subprojects/docs/src/samples/userguide/dependencies/iteratingArtifacts/build.gradle]: done

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc [hunk 6]
```diff
@@ -15,73 +15,47 @@
 [[working_with_dependencies]]
 == Working with Dependencies
 
-For the examples below we have the following dependencies setup:
+Gradle provides an extensive API for navigating, inspecting and post-processing dependencies and artifacts of resolved modules. With this API at hand, you can implement more complex requirements to your build requiring a "runtime view" of dependencies.
```
[the remark this answers is not in the document]
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: done

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc (hunk 2 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: I'll remove it for now. In the beginning I tried to pin back to the content in "Inspecting Dependencies" but later I decided to keep the content in "Working with Dependencies".

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc (hunk 3 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: done

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc (hunk 4 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: I see, ok. done

File: subprojects/docs/src/docs/userguide/workingWithDependencies.adoc (hunk 5 above)
Comment (bmuschko) [reply subprojects/docs/src/docs/userguide/workingWithDependencies.adoc]: > "Gradle downloads the main artifact and the metadata file for a module and stores them" -> "Gradle downloads the metadata file of a module and stores it"

done

> The part after the first sentence starting with The initial download... feels unrelated to this section.

I agree. Works better without. Removed.

**Comment by bmuschko** [comment]
> There are more related APIs though. But honestly I don't have an overview of what might be explained elsewhere already.

This is where it get kind over blurry for me. We have some overlap between APIs accessible from `Configuration` (same or similar information but via a different method). It's not clear to me when to use what. I think if we can come up with concrete use cases then we expand this section. Happy to hear your input.

**Pull request merged**